src/arm/dynarec-arm/dynarec-impl.c

```diff
@@ -0,0 +1,1225 @@
+DEFINE_LOAD_STORE_WITH_REGISTER_THUMB_LOAD(LDRB2, cpu->memory.load8, EMIT(ctx, MOV, AL, reg_rd, 0))
+DEFINE_LOAD_STORE_WITH_REGISTER_THUMB_LOAD(LDRH2, cpu->memory.load16, EMIT(ctx, MOV, AL, reg_rd, 0))
+DEFINE_LOAD_STORE_WITH_REGISTER_THUMB_LOAD(LDRSB, cpu->memory.load8, EMIT(ctx, SXTB, AL, reg_rd, 0, 0))
+
+DEFINE_LOAD_STORE_WITH_REGISTER_THUMB(LDRSH,
+	// TODO(merry): Improve
+	// Interpreter:
+	//     rm = cpu->gprs[rn] + cpu->gprs[rm];
+	//     cpu->gprs[rd] = rm & 1
+	//                     ? ARM_SXT_8(cpu->memory.load16(cpu, rm, &currentCycles))
+	//                     : ARM_SXT_16(cpu->memory.load16(cpu, rm, &currentCycles));
+	FLUSH_HOST_NZCV
+	PREPARE_FOR_BL
+	flushPC(ctx);
+	flushPrefetch(ctx);
+	loadRegValue(ctx, rn, 1);
+	loadRegValue(ctx, rm, 2);
+	EMIT(ctx, ADD, AL, 1, 1, 2);
+	EMIT(ctx, ADDI, AL, 2, REG_ARMCore, offsetof(struct ARMCore, cycles));
+	PUSH_REGLIST_SAVE;
+	EMIT(ctx, BL, AL, ctx->code, cpu->memory.load16);
+	// TODO(merry): Is there an alternative to remateralizing here?
+	EMIT(ctx, MOV, AL, 3, 0);
+	POP_REGLIST_SAVE;
+	loadRegValue(ctx, rn, 1);
+	loadRegValue(ctx, rm, 2);
+	EMIT(ctx, ADD, AL, 1, 1, 2);
+	EMIT(ctx, TSTI, AL, 1, 1);
+	unsigned reg_rd = defReg(ctx, rd);
+	EMIT(ctx, SXTB, NE, reg_rd, 3, 0);
+	EMIT(ctx, SXTH, EQ, reg_rd, 3, 0);
+	saveRegs(ctx);
+	THUMB_LOAD_POST_BODY;)
+
+#define DEFINE_LOAD_STORE_WITH_REGISTER_THUMB_STORE(NAME, FUNC) \
+	DEFINE_LOAD_STORE_WITH_REGISTER_THUMB(NAME, \
+		FLUSH_HOST_NZCV \
+		PREPARE_FOR_BL \
+		flushPC(ctx); \
+		flushPrefetch(ctx); \
+		loadRegValue(ctx, rn, 1); \
+		loadRegValue(ctx, rm, 3); \
+		EMIT(ctx, ADD, AL, 1, 1, 3); \
+		loadRegValue(ctx, rd, 2); \
+		EMIT(ctx, ADDI, AL, 3, REG_ARMCore, offsetof(struct ARMCore, cycles)); \
+		PUSH_REGLIST_SAVE; \
+		EMIT(ctx, BL, AL, ctx->code, FUNC); \
+		POP_REGLIST_SAVE; \
+		saveRegs(ctx); \
+		THUMB_STORE_POST_BODY;)
+
+DEFINE_LOAD_STORE_WITH_REGISTER_THUMB_STORE(STR2, cpu->memory.store32)
+DEFINE_LOAD_STORE_WITH_REGISTER_THUMB_STORE(STRB2, cpu->memory.store8)
+DEFINE_LOAD_STORE_WITH_REGISTER_THUMB_STORE(STRH2, cpu->memory.store16)
+
+#define DEFINE_LOAD_STORE_MULTIPLE_THUMB(NAME, RN, LS, DIRECTION, PRE_BODY, WRITEBACK) \
+	DEFINE_INSTRUCTION_THUMB(NAME, \
+		FLUSH_HOST_NZCV \
+		PREPARE_FOR_BL \
+		flushPC(ctx); \
+		flushPrefetch(ctx); \
+		flushRegisterCache(ctx); \
+		int rn = RN; \
+		int rs = opcode & 0xFF; \
+		PRE_BODY; \
+		loadRegValue(ctx, rn, 1); \
+		EMIT_IMM(ctx, AL, 2, rs); \
+		PUSH_REGLIST_SAVE; /* needs to be here due to stack manipulation */  \
+		EMIT(ctx, SUBI, AL, ARM_SP, ARM_SP, 8); \
+		/* here we load the &cycles argument: */ \
+		EMIT(ctx, ADDI, AL, 3, REG_ARMCore, offsetof(struct ARMCore, cycles)); \
+		EMIT(ctx, STRI, AL, 3, ARM_SP, 0); \
+		/* arguments: (cpu, address, rs, LSM_ ## DIRECTION, &cycles) */ \
+		EMIT_IMM(ctx, AL, 3, LSM_ ## DIRECTION); \
+		EMIT(ctx, BL, AL, ctx->code, cpu->memory.CONCAT2(LS, Multiple)); \
```

Comment:
Why is CONCAT2 needed? See `src/arm/debugger/memory-shim.c` for how I do this without a macro.